core/stat/base/leap_array.go

```diff
@@ -0,0 +1,221 @@
+// For example, windowLengthInMs is 500ms, intervalInMs is 1min, so sampleCount is 120.
+type leapArray struct {
+	windowLengthInMs uint32
+	sampleCount      uint32
+	intervalInMs     uint32
+	array            *atomicWindowWrapArray
+	// update lock
+	updateLock triableMutex
+}
+
+func (la *leapArray) currentWindow(bg bucketGenerator) (*windowWrap, error) {
+	return la.currentWindowWithTime(util.CurrentTimeMillis(), bg)
+}
+
+func (la *leapArray) currentWindowWithTime(timeMillis uint64, bg bucketGenerator) (*windowWrap, error) {
+	if timeMillis < 0 {
+		return nil, errors.New("Current time is less than 0.")
+	}
+
+	idx := la.calculateTimeIdx(timeMillis)
+	windowStart := calculateStartTime(timeMillis, la.windowLengthInMs)
+
+	for { //spin to get the current windowWrap
+		old := la.array.get(idx)
+		if old == nil {
+			// because la.array.data had initiated when new la.array
+			// theoretically, here is not reachable
+			newWrap := &windowWrap{
+				windowStart: windowStart,
+				value:       bg.newEmptyBucket(),
+			}
+			if la.array.compareAndSet(idx, nil, newWrap) {
+				return newWrap, nil
+			} else {
+				runtime.Gosched()
+			}
+		} else if windowStart == atomic.LoadUint64(&old.windowStart) {
+			return old, nil
+		} else if windowStart > atomic.LoadUint64(&old.windowStart) {
+			// current time has been next cycle of leapArray and leapArray dont't count in last cycle.
+			// reset windowWrap
+			if la.updateLock.TryLock() {
+				old = bg.resetWindowTo(old, windowStart)
+				la.updateLock.Unlock()
+				return old, nil
+			} else {
+				runtime.Gosched()
+			}
+		} else if windowStart < old.windowStart {
+			// used for some special case(e.g. when occupying "future" buckets).
+			return nil, errors.New(fmt.Sprintf("Provided time timeMillis=%d is already behind old.windowStart=%d.", windowStart, old.windowStart))
+		}
+	}
+}
+
+func (la *leapArray) calculateTimeIdx(timeMillis uint64) int {
+	timeId := timeMillis / uint64(la.windowLengthInMs)
+	return int(timeId) % la.array.length
+}
+
+//  Get all windowWrap between [current time -1000ms, current time]
+func (la *leapArray) values() []*windowWrap {
+	return la.valuesWithTime(util.CurrentTimeMillis())
+}
+
+func (la *leapArray) valuesWithTime(timeMillis uint64) []*windowWrap {
+	if timeMillis <= 0 {
+		return make([]*windowWrap, 0)
+	}
+	ret := make([]*windowWrap, 0)
+	for i := 0; i < la.array.length; i++ {
+		ww := la.array.get(i)
+		if ww == nil || la.isWindowDeprecated(timeMillis, ww) {
+			continue
+		}
+		newWW := &windowWrap{
+			windowStart: ww.windowStart,
+			value:       ww.value,
+		}
+		ret = append(ret, newWW)
+	}
+	return ret
+}
+
+func (la *leapArray)ValuesWithConditional(timeMillis uint64, predicate base.TimePredicate) []*windowWrap {
+	if timeMillis <= 0 {
+		return make([]*windowWrap, 0)
+	}
+	ret := make([]*windowWrap, 0)
+	for i := 0; i < la.array.length; i++ {
+		ww := la.array.get(i)
+		if ww == nil || la.isWindowDeprecated(timeMillis, ww) || !predicate(timeMillis) {
```

Comment:
This should be the predicate for the startTime of the bucket, rather than current time? We leverage the bucketStartTime to check whether the bucket satisfy the condition.